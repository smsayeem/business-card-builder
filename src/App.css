/* google font */
@import url('https://fonts.googleapis.com/css2?family=Tinos:wght@400;700&display=swap');

/* color variables */
:root {
  --primaryColor: rgb(61,79,97);
  --primaryGrey: rgba(61,79,97, 0.8);
  --primaryLight: rgba(61,79,97, 0.4);

  --primaryBtn: rgb(93,168,221);
  --primaryBtnDark: rgb(30, 112, 172);
  --secondaryBtn: rgb(122,141,155);
  --secondaryBtnDark: rgb(61,79,97);

  --mainWhite: rgb(255,255,255);
  --mainGrey: rgb(231,233,236);
  --mainBlack: rgb(50,50,50);
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

body {
  background: var(--mainBlack);
  color: var(--primaryColor);
  font-family: 'Tinos', serif;
}

/* remove Arrows/Spinners from the input number fields */
input[type=number]::-webkit-inner-spin-button, 
input[type=number]::-webkit-outer-spin-button { 
  -webkit-appearance: none; 
  margin: 0; 
}

.App {
  max-width: 1160px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  background-color: var(--mainGrey);
}

@media screen and (min-width: 769px){
  .App {
    flex-direction: row;
  }
}
