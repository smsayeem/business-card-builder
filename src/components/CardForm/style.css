
.form-container{
  display: flex;
  flex-direction: column;
  padding: 3rem 1rem 1rem 1rem;
  background-color: var(--mainWhite);
  width: 100%;
}

header h1{
  letter-spacing: 1px;
  font-size: 2rem;
}

.divider{
  border-bottom: 1px solid var(--mainGrey);
  margin: 2rem 0 1rem;
}
.divider p{
  text-transform: uppercase;
  padding-bottom: 6px;
  font-size: 0.8rem;
  color: var(--primaryLight);
}

.form-center {
  display: flex;
  flex-wrap: wrap;
  flex: 0 0 100%;
}

.form-group {
  flex: 0 0 100%;
}

label {
  display: block;
  color: var(--primaryGrey);
  font-size: 0.8rem;
  text-transform: capitalize;
  margin: 1rem 0 0.2rem;
  text-transform: uppercase;
}

.form-input {
  border: 1px solid var(--primaryLight);
  border-radius: 4px;
  outline: none;
  width: 100%;
  font-size: 1rem;
  padding: 10px;
  color: var(--mainBlack);
  letter-spacing: 1px;
}

.btn-group {
  padding: 1rem 0;
  margin: 1rem 0 0 0;
  width: 100%;
  display: flex;
  justify-content: space-between;
}

.btn {
  width: 48%;
  height: 2.7rem;
  text-decoration: none;
  color: var(--mainWhite);
  background-color: var(--mainBlack);
  text-align: center;
  text-transform: capitalize;
  letter-spacing: 1px;
  font-size: 1rem;
  outline: 0;
  border: none;
  border-radius: 4px;
  padding: 0 1rem;
  cursor: pointer;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.2);
  transition: all 0.3s ease;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  -ms-transition: all 0.3s ease;
  -o-transition: all 0.3s ease;
}
.btn:hover { box-shadow: 0 3px 3px rgba(0, 0, 0, 0.3); }
.btn-avatar{ background-color: var(--secondaryBtn); }
.btn-avatar:hover{ background-color: var(--secondaryBtnDark); }
.btn-hcard{ background-color: var(--primaryBtn); }
.btn-hcard:hover{ background-color: var(--primaryBtnDark); }

@media screen and (min-width: 430px){
  .form-container{ padding: 3rem; }
  .form-group { flex: 0 0 50%; }
  .form-group:nth-child(odd) { padding-right: 8px; }
  .form-group:nth-child(even) { padding-left: 8px; }
  .btn-group { margin: 1rem 0 0 0; }
  .btn{ font-size: 1.1rem; }
}

@media screen and (min-width: 769px){
  .form-container{ flex: 0 0 50%; }
}
