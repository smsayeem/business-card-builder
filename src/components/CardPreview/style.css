.preview{
  max-width: 375px;
  width: 100%;
  margin: 0px auto;
  padding: 2rem 0 3rem;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.card-title{
  margin: 0;
  font-size: 1rem;
  text-align: right;
  font-weight: 400;
  text-transform: uppercase;
  color: var(--primaryLight);
  padding-bottom: 6px;
}
.card-header{
  height: 88px;
  background-color: rgb(62,79,97);
  color: #fff;
  text-transform: capitalize;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  padding: 0.8rem;
  position: relative;
}

.card-header h2{
  font-size: 1.5rem;
  text-transform: capitalize;
  margin:0;
}

.card-avatar{
  position: absolute;
  right: 8px;
  top: 8px;
  width: 70px;
  height: 89px;
  background-color: var(--mainWhite);
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.2);
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
  object-fit: cover;
}

.card-avatar img{
  padding: 6px;
  max-width: 100%;
  max-height: 100%;
}

.card-body{
  padding: 24px 12px 18px;
  background-color: #fff;
  border-bottom: 2px solid var(--primaryLight);
}

.card-items{
  display: flex;
  border-bottom: 1px solid var(--mainGrey);
  justify-content: flex-end;
  margin-bottom: 10px;
  padding-bottom: 2px;
}

.label{
  flex: 0 0 22%;
  text-transform: uppercase;
  color: rgb(62,79,97);
  font-size: 0.6rem;
}

.info{
  height: 1rem;
  flex: 0 0 78%;
  font-size: 0.8rem;
  letter-spacing: 0.2px;
  color: var(--mainBlack);
}

.card-special-item-wrapper{
  display: flex;
}

.card-special-items{
  flex: 0 0 50%;
  display: flex;
  padding-bottom: 2px;
  border-bottom: 1px solid var(--mainGrey);
}

.info-special{
  height: 1rem;
  margin-left: 16px;
  font-size: .8rem;
  letter-spacing: .2px;
  color: var(--mainBlack);
}

@media screen and (min-width: 769px){
  .label{ font-size: 0.7rem }
  .info{
    font-size: 0.9rem;
    letter-spacing: 1px;
  } 
  .info-special{
    margin-left: 24px;
    font-size: 0.9rem;
    letter-spacing: .2px;
  }
}

@media screen and (min-width: 1025px){
  .preview{ min-width: 426px; }
  .card-header{  
    height: 100px;
    padding: 0.9rem;
  }
  .card-header h2{ font-size: 1.7rem; }
  .card-avatar{
    position: absolute;
    right: 12px;
    top: 12px;
    width: 82px;
    height: 104px;
  }
  .card-body{ padding: 26px 18px 22px; }
  .label{ font-size: .8rem; }
  .info{
    font-size: 1rem;
    letter-spacing: 1px;
  }
  .info-special{
    margin-left: 24px;
    font-size: 1rem;
    letter-spacing: 1px;
  }
}