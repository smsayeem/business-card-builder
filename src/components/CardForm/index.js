import React from 'react';
import PropTypes from 'prop-types';
import FormElement from './components/FormElement';
import isvalidForm from '../../utilities/helper/isvalidForm';
import Divider from './components/Divider'

import './style.css';

const CardForm = ({ handleChange, handleSubmit, handleImage }) => {

  // when 'Upload Avatar' button is clicked this function targets the input ('#upload-button') and add a callback function to upload image
  const clickUpload=()=>{
    const uploadBtn = document.querySelector('#upload-button');
    uploadBtn.click();
  }

  const validateFormData = (e) => {
    e.preventDefault();
    isvalidForm(e) && handleSubmit(e);
  }
  return (
    <div className="form-container">
      <header>
        <h1>hCard Builder</h1>
      </header>
      <form onSubmit={validateFormData}>
        <Divider text="personal details" />
        <section className="form-center">
          <FormElement formField="input" handleChange={handleChange} name="givenname" title="given name" type="text" />
          <FormElement formField="input" handleChange={handleChange} name="surname" title="surname" type="text" />
          <FormElement formField="input" handleChange={handleChange} name="email" title="email" type="email" />
          <FormElement formField="input" handleChange={handleChange} name="phone" title="phone" type="text" />
        </section>
        <Divider text="personal details" />
        <section className="form-center">
          <FormElement formField="input" handleChange={handleChange} name="houseno" title="houseno" type="text" />
          <FormElement formField="input" handleChange={handleChange} name="street" title="street" type="text" />
          <FormElement formField="input" handleChange={handleChange} name="suburb" title="suburb" type="text" />
          <FormElement formField="select" handleChange={handleChange} name="state" title="state" type="text" />
          <FormElement formField="input" handleChange={handleChange} name="postcode" title="postcode" type="number" />
          <FormElement formField="input" handleChange={handleChange} name="country" title="country" type="text" />
          {/* buttons */}
          <div className="btn-group ">
            {/* this input allows to select image files. It is active but hidden as per design  */}
            <input name="img-upload" id="upload-button" hidden type='file' onChange={handleImage} />

            {/* this button will call a function 'clickUpload' that will trigger the input above to upload an image */}
            <input readOnly name="upload" id="btn-avatar" className = "btn btn-avatar" onClick = {clickUpload}  value="Upload Avatar" />
            
            {/* no functionality added for this button as it is not part of the test */}
            <button type="submit" name="submit" className = "btn btn-hcard">
              Create hCard
            </button>
          </div>
        </section>
      </form>
    </div>
  )
}

CardForm.propTypes = {
  handleChange: PropTypes.func, 
  handleSubmit: PropTypes.func, 
  handleImage: PropTypes.func,
};

CardForm.defaultProps = {
  handleChange: ()=>{}, 
  handleSubmit: ()=>{}, 
  handleImage: ()=>{},
};

export default CardForm
