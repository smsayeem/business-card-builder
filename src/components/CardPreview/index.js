import React, { Component } from 'react';
import './style.css';

export class CardPreview extends Component {
  render() {
    return (
      <div>
      CardPreview
      </div>
    )
  }
}

export default CardPreview
