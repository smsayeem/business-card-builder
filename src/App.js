import React, { useState } from 'react';
import { Helmet } from 'react-helmet';
import CardPreview from './components/CardPreview';
import CardForm from './components/CardForm';
import './App.css';

function App() {

  // updating user's data that coming from the form fields
  const [ user, setUser ] = useState({});

  // keep image url that is uploaded via 'Upload Avatar' button click in the form
  const [ image, setImage ] = useState({ preview_imgUrl: '' });

  const handleImage = (e) => {
    // 'preview_imgUrl' is used for screen display. 'URL.createObjectURL(obj)' creates a DOMString containing a URL representing the object given in the parameter
    setImage({
     preview_imgUrl: URL.createObjectURL(e.target.files[0]),
    })
   }

  const handleChange = (e) => {
    setUser({ ...user, [e.target.name]:e.target.value });
  }

  // No functionality added for the submit button.
  const handleSubmit = (e) => {
    alert('congratulation !!! Form submitted successfully.')
  }

  return (
    <section className = 'App'>
      <Helmet>
        <title>hCard  Builder</title>
        <meta name = "description" content = "hCard Builder is a simple app, publishing people, companies and organizations on the web" />
      </Helmet>
      <CardForm handleChange = { handleChange } handleSubmit = { handleSubmit } handleImage = { handleImage } />
      <CardPreview user={user} avatar={image} />
    </section>
  );
}

export default App;
