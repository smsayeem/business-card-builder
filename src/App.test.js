import React from 'react';
import { shallow, mount } from 'enzyme';

import App from './App';
import CardForm from './components/CardForm';
import CardPreview from './components/CardPreview';
import logoImage from './components/assets/sample_fallback.png';


const userData={
  user:{
    givenname: 'sam', surname: 'sadot', email: 'sam@sam.com', phone:'0433530825', houseno:'1a', street:'King street', suburb: 'Sydney', state: 'NSW', postcode:'2000', country: 'Australia'
  },
  avatar:{
    preview_imgUrl: logoImage
  }
}
describe('App and other components existence check', () => {
  let wrapper;
    beforeEach(()=>{
      wrapper = mount(<App />);
  });

  // snapshot test
  it('should render the <App /> Component correctly', () => {   
    expect(wrapper).toMatchSnapshot();
  });

  // render the App componnet without crashing
  it('should render the App componnet without crashing', ()=>{
    shallow(<App />);
  });
  // render the CardForm componnet without crashing
  it('should render the CardForm componnet without crashing', ()=>{
    shallow(<CardForm />);
  });
      // render the CardPreview componnet without crashing
  it('should render the CardPreview componnet without crashing', ()=>{
    shallow(<CardPreview user={userData.user} avatar={userData.avatar} />);
  });

  it('should render the CardForm and CardPreview components', ()=>{
    const cardForm = wrapper.find(CardForm)
    const cardPreview = wrapper.find(CardPreview)

    expect(cardForm.exists()).toBe(true);
    expect(cardPreview.exists()).toBe(true);
  })
});
describe('passing props', ()=>{
  const cardWrapper = mount(<CardPreview user={userData.user} avatar={userData.avatar} />);
  it('accept the CardPreview props', ()=>{
    expect(cardWrapper.props().user).toEqual(userData.user);
    expect(cardWrapper.props().avatar).toEqual(userData.avatar);
  })
})
describe('test component logic', ()=>{
  const handleChange = jest.fn();
  const handleSubmit = jest.fn();
  const handleImage = jest.fn();

  const cardWrapper = mount(<CardPreview user={userData.user} avatar={userData.avatar} />);
  const formWrapper = mount(<CardForm handleChange = { handleChange} handleSubmit = { handleSubmit } handleImage = { handleImage } />);
  const img_upload = formWrapper.find('#upload-button');

  it('button click - upload image', ()=>{
    img_upload.simulate('click');
    img_upload.simulate('change');
    const img_url = cardWrapper.find('img').prop("src");
    expect(img_url).toEqual(logoImage);
  })
})