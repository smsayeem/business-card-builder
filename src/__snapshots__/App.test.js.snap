// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App and other components existence check should render the <App /> Component correctly 1`] = `
<App>
  <section
    className="App"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        defer={true}
        encodeSpecialCharacters={true}
        meta={
          Array [
            Object {
              "content": "hCard Builder is a simple app, publishing people, companies and organizations on the web",
              "name": "description",
            },
          ]
        }
        title="hCard  Builder"
        titleAttributes={Object {}}
      >
        <NullComponent
          defer={true}
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "hCard Builder is a simple app, publishing people, companies and organizations on the web",
                "name": "description",
              },
            ]
          }
          title="hCard  Builder"
          titleAttributes={Object {}}
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
    <CardForm
      handleChange={[Function]}
      handleImage={[Function]}
      handleSubmit={[Function]}
    >
      <div
        className="form-container"
      >
        <header>
          <h1>
            hCard Builder
          </h1>
        </header>
        <form
          onSubmit={[Function]}
        >
          <Divider
            text="personal details"
          >
            <div
              className="divider"
            >
              <p>
                personal details
              </p>
            </div>
          </Divider>
          <section
            className="form-center"
          >
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="givenname"
              title="given name"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="givenname"
                >
                  given name
                </label>
                <input
                  className="form-input"
                  name="givenname"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="surname"
              title="surname"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="surname"
                >
                  surname
                </label>
                <input
                  className="form-input"
                  name="surname"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="email"
              title="email"
              type="email"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="email"
                >
                  email
                </label>
                <input
                  className="form-input"
                  name="email"
                  onChange={[Function]}
                  type="email"
                />
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="phone"
              title="phone"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="phone"
                >
                  phone
                </label>
                <input
                  className="form-input"
                  name="phone"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
          </section>
          <Divider
            text="personal details"
          >
            <div
              className="divider"
            >
              <p>
                personal details
              </p>
            </div>
          </Divider>
          <section
            className="form-center"
          >
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="houseno"
              title="houseno"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="houseno"
                >
                  houseno
                </label>
                <input
                  className="form-input"
                  name="houseno"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="street"
              title="street"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="street"
                >
                  street
                </label>
                <input
                  className="form-input"
                  name="street"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="suburb"
              title="suburb"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="suburb"
                >
                  suburb
                </label>
                <input
                  className="form-input"
                  name="suburb"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
            <FormElement
              formField="select"
              handleChange={[Function]}
              name="state"
              title="state"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="state"
                >
                  state
                </label>
                <select
                  className="form-input"
                  name="state"
                  onChange={[Function]}
                >
                  <option
                    value="-1"
                  >
                    Select your state
                  </option>
                  <option
                    value="NSW"
                  >
                    NSW
                  </option>
                  <option
                    value="ACT"
                  >
                    ACT
                  </option>
                  <option
                    value="VIC"
                  >
                    VIC
                  </option>
                  <option
                    value="QLD"
                  >
                    QLD
                  </option>
                  <option
                    value="TAS"
                  >
                    TAS
                  </option>
                  <option
                    value="WA"
                  >
                    WA
                  </option>
                </select>
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="postcode"
              title="postcode"
              type="number"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="postcode"
                >
                  postcode
                </label>
                <input
                  className="form-input"
                  name="postcode"
                  onChange={[Function]}
                  type="number"
                />
              </div>
            </FormElement>
            <FormElement
              formField="input"
              handleChange={[Function]}
              name="country"
              title="country"
              type="text"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="country"
                >
                  country
                </label>
                <input
                  className="form-input"
                  name="country"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </FormElement>
            <div
              className="btn-group "
            >
              <input
                hidden={true}
                id="upload-button"
                name="img-upload"
                onChange={[Function]}
                type="file"
              />
              <input
                className="btn btn-avatar"
                id="btn-avatar"
                name="upload"
                onClick={[Function]}
                readOnly={true}
                value="Upload Avatar"
              />
              <button
                className="btn btn-hcard"
                name="submit"
                type="submit"
              >
                Create hCard
              </button>
            </div>
          </section>
        </form>
      </div>
    </CardForm>
    <CardPreview
      avatar={
        Object {
          "preview_imgUrl": "",
        }
      }
      user={Object {}}
    >
      <section
        className="preview"
      >
        <h1
          className="card-title"
        >
          hCard preview
        </h1>
        <header
          className="card-header"
        >
          <h2>
             
          </h2>
          <div
            className="card-avatar"
          >
            <img
              alt="avatar"
              src="sample_fallback.png"
            />
          </div>
        </header>
        <section
          className="card-body"
        >
          <div
            className="card-items"
          >
            <div
              className="label"
            >
              email
            </div>
            <div
              className="info"
            />
          </div>
          <div
            className="card-items"
          >
            <div
              className="label"
            >
              phone
            </div>
            <div
              className="info"
            />
          </div>
          <div
            className="card-items"
          >
            <div
              className="label"
            >
              address
            </div>
            <div
              className="info"
            >
               
            </div>
          </div>
          <div
            className="card-items"
          >
            <div
              className="label"
              style={
                Object {
                  "visibility": "hidden",
                }
              }
            >
              address2
            </div>
            <div
              className="info"
            >
               
            </div>
          </div>
          <div
            className="card-special-item-wrapper"
          >
            <div
              className="card-special-items"
            >
              <div
                className="label"
              >
                postcode
              </div>
              <div
                className="info-special"
              />
            </div>
            <div
              className="card-special-items"
            >
              <div
                className="label"
              >
                country
              </div>
              <div
                className="info-special"
              />
            </div>
          </div>
        </section>
      </section>
    </CardPreview>
  </section>
</App>
`;
